File: ui/imports/api/scans/scans.js
/////////////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2017 Koren Lev (Cisco Systems), Yaron Yogev (Cisco Systems) and others /
//                                                                                      /
// All rights reserved. This program and the accompanying materials                     /
// are made available under the terms of the Apache License, Version 2.0                /
// which accompanies this distribution, and is available at                             /
// http://www.apache.org/licenses/LICENSE-2.0                                           /
/////////////////////////////////////////////////////////////////////////////////////////
import { Mongo } from 'meteor/mongo';
import { SimpleSchema } from 'meteor/aldeed:simple-schema';
import * as R from 'ramda';

import { Constants } from '/imports/api/constants/constants';
import { StatusesInOperation } from '/imports/api/constants/data/scans-statuses';

export const Scans = new Mongo.Collection('scans', { idGeneration: 'MONGO' });

Scans.schemaRelated = {
  environment: {
    label: 'Environment',
    description: 'Name of environment to scan',
    disabled: true,
  },
  status: {
    label: 'Status',
    description: 'Scan lifecycle status',
    subtype: 'select',
    options: 'scans_statuses',
    disabled: true,
  },
  object_id: {
    label: 'Scan specific object',
    description: 'Object ID',
  },
  log_level: {
    label: 'Log level',
    description: 'logging level',
    subtype: 'select',
    options: 'log_levels',
  },
  clear: {
    label: 'Clear data',
    description: 'clear all data prior to scanning',
  },
  scan_only_inventory: {
    label: 'Scan only inventory',
    description: 'do only scan to inventory',
  },
  scan_only_links: {
    label: 'Scan only links',
    description: 'do only links creation',
  },
  scan_only_cliques: {
    label: 'Scan only cliques',
    description: 'do only cliques creation',
  },
};

Scans.scansOnlyFields = ['scan_only_inventory', 'scan_only_links', 'scan_only_cliques'];

let schema = {
  _id: { type: { _str: { type: String, regEx: SimpleSchema.RegEx.Id } } },
  environment: { 
    type: String 
  }, 
  status: {
    type: String, 
    defaultValue: 'draft', 
    custom: function () {
      let that = this;
      let statuses = Constants.findOne({ name: 'scans_statuses' }).data;

      if (R.isNil(R.find(R.propEq('value', that.value), statuses))) {
        return 'notAllowed';
      }
    },
  },
  object_id: {
    type: String,
    optional: true,
  },
  log_level: {
    type: String,
    defaultValue: 'warning', 
    custom: function () {
      let that = this;
      let logLevels = Constants.findOne({ name: 'log_levels' }).data;

      if (R.isNil(R.find(R.propEq('value', that.value), logLevels))) {
        return 'notAllowed';
      }
    },
  },
  clear: {
    type: Boolean,
    defaultValue: true, 
  },
  scan_only_inventory: {
    type: Boolean,
    defaultValue: false,
  },
  scan_only_links: {
    type: Boolean,
    defaultValue: false,
  },
  scan_only_cliques: {
    type: Boolean,
    defaultValue: false,
  },
  submit_timestamp: {
    type: Date,
    defaultValue: null
  },

};

Scans.schema = new SimpleSchema(schema);
Scans.schema.addValidator(function () {
  let that = this;
  let env = that.field('environment').value;

  let currentScansCount = Scans.find({
    environment: env,
    status: { $in: StatusesInOperation }
  }).count();

  if (currentScansCount > 0) {
    throw {
      isError: true,
      type: 'notUinque',
      data: [],
      message: 'There is already a scan in progress.'
    };
  }

  let scanOnlyFields = R.filter( f => that.field(f).value, Scans.scansOnlyFields);

  if(scanOnlyFields.length > 1) {
    throw {
      isError: true,
      type: 'conflict',
      data: scanOnlyFields,
      message: 'Only one of the scan only fields can be selected'
    };
  }

});

Scans.attachSchema(Scans.schema);

Scans.schemaRelated = R.mapObjIndexed((relatedItem, key) => {
  return R.merge(relatedItem, {
    type: schema[key].type 
  });

}, Scans.schemaRelated);

export const subsScansEnvPageAmountSorted = 'scans?env*&page&amount&sortField&sortDirection';
export const subsScansEnvPageAmountSortedCounter = `${subsScansEnvPageAmountSorted}!counter`;

File: app/test/api/responders_test/test_data/scans.py
###############################################################################
# Copyright (c) 2017 Koren Lev (Cisco Systems), Yaron Yogev (Cisco Systems)   #
# and others                                                                  #
#                                                                             #
# All rights reserved. This program and the accompanying materials            #
# are made available under the terms of the Apache License, Version 2.0       #
# which accompanies this distribution, and is available at                    #
# http://www.apache.org/licenses/LICENSE-2.0                                  #
###############################################################################
from test.api.responders_test.test_data import base

URL = "/scans"

WRONG_ID = base.WRONG_OBJECT_ID
NONEXISTENT_ID = "58c96a075eb66a121cc4e750"
CORRECT_ID = base.CORRECT_OBJECT_ID

BASE_OBJECT = "node-2.cisco.com"

WRONG_STATUS = base.WRONG_SCAN_STATUS
CORRECT_STATUS = base.CORRECT_SCAN_STATUS

SCANS = [
    {
        "status": "pending",
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75f",
    },
    {
        "status": "completed",
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75e",
        "scan_completed": True
    }
]

SCANS_RESPONSE = {
    "scans": SCANS
}

SCANS_WITH_SPECIFIC_ID = [
    {
        "status": "pending",
        "environment": "Mirantis-Liberty-API",
        "id": CORRECT_ID,
    }
]

SCANS_WITH_SPECIFIC_BASE_OBJ = [
    {
        "status": "pending",
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75f",
        "object_id": BASE_OBJECT
    },
    {
        "status": "completed",
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75e",
        "object_id": BASE_OBJECT,
        "scan_completed": True
    }
]

SCANS_WITH_SPECIFIC_BASE_OBJ_RESPONSE = {
    "scans": SCANS_WITH_SPECIFIC_BASE_OBJ
}

SCANS_WITH_SPECIFIC_STATUS = [
    {
        "status": CORRECT_STATUS,
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75f",
        "scan_completed": True
    },
    {
        "status": CORRECT_STATUS,
        "environment": "Mirantis-Liberty-API",
        "id": "58c96a075eb66a121cc4e75e",
        "scan_completed": True
    }
]

SCANS_WITH_SPECIFIC_STATUS_RESPONSE = {
    "scans": SCANS_WITH_SPECIFIC_STATUS
}

NON_DICT_SCAN = base.NON_DICT_OBJ

SCAN = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": True,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITHOUT_ENV = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": True,
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_UNKNOWN_ENV = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": True,
       "environment": base.UNKNOWN_ENV,
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITHOUT_STATUS = {
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": True,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_WRONG_STATUS = {
       "status": WRONG_STATUS,
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": True,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_WRONG_LOG_LEVEL = {
       "status": "pending",
       "log_level": base.WRONG_LOG_LEVEL,
       "clear": True,
       "scan_only_inventory": True,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_NON_BOOL_CLEAR = {
       "status": "pending",
       "log_level": "warning",
       "clear": base.NON_BOOL,
       "scan_only_inventory": True,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}


SCAN_WITH_NON_BOOL_SCAN_ONLY_INVENTORY = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_inventory": base.NON_BOOL,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_NON_BOOL_SCAN_ONLY_LINKS = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_links": base.NON_BOOL,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}

SCAN_WITH_NON_BOOL_SCAN_ONLY_CLIQUES = {
       "status": "pending",
       "log_level": "warning",
       "clear": True,
       "scan_only_cliques": base.NON_BOOL,
       "environment": "Mirantis-Liberty-API",
       "inventory": "inventory",
       "object_id": "ff"
}
